tests/tests_pytorch/strategies/test_sharded_strategy.py

```diff
@@ -18,6 +20,37 @@
     from fairscale.optim import OSS
 
 
+class ModelWithAdamOptimizer(BoringModel):
+    def configure_optimizers(self):
+        optimizer = torch.optim.Adam(self.layer.parameters(), lr=0.1)
+        return optimizer
+
+
+class CheckModelRestore(ModelWithAdamOptimizer):
+    def __init__(self, old_model_state_dict, old_optimizer_states):
+        super().__init__()
+        self.old_model_state_dict = old_model_state_dict
+        self.old_optimizer_states = old_optimizer_states
+
+    def on_train_start(self):
+        assert all(
+            self._is_equal(actual, expected) for actual, expected in zip(self.state_dict(), self.old_model_state_dict)
+        )
+
+        for optimizer, state in zip(self.trainer.optimizers, self.old_optimizer_states):
+            optimizer.consolidate_state_dict()
+            self._is_equal(optimizer.state_dict(), state)
```

Comment:
I do not understand why this is necessary. 

When we checkpoint the states, we already consolidate the state dict. When the state gets loaded for the single device, it already is consolidated. And optimizer.consolidate_state_dict should be a no-op anyway on single device. 

Please remind me where I am wrong here.

Reply:
looks like one can call `state_dict(all_ranks=True)` to do that.
https://github.com/facebookresearch/fairscale/blob/ee647b976cf4c8fdd37bc9ae3fd6331d225ba2a0/fairscale/optim/oss.py#L400-L404

since it relies on `self._all_states` with is an empty list by default.

Reply:
No, there must be a misunderstanding here. You set devices=1, there is only one rank. [consolidation](https://pytorch.org/docs/stable/distributed.optim.html#torch.distributed.optim.ZeroRedundancyOptimizer.consolidate_state_dict) is essentially a no-op here. Can you explain why you did it?

Reply:
yes, I understand that theoretically, consolidate call is not required here but that's not how fairscale is designed. it relies on `self._all_states` to generate the `state_dict`
https://github.com/facebookresearch/fairscale/blob/ee647b976cf4c8fdd37bc9ae3fd6331d225ba2a0/fairscale/optim/oss.py#L305